clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -184,16 +184,22 @@ private void startSequencesAtBeginning(TopicPartition topicPartition, ProducerId
         // responses which are due to the retention period elapsing, and those which are due to actual lost data.
         private long lastAckedOffset;
 
+        private static final Comparator<ProducerBatch> PRODUCER_BATCH_COMPARATOR = (b1, b2) -> {
+            if (b1.baseSequence() < b2.baseSequence()) return -1;
+            else if (b1.baseSequence() > b2.baseSequence()) return 1;
+            else return b1.equals(b2) ? 0 : 1;
```

Comment:
Hi @artemlivshits, thanks for your comment. I don't think it violate the requirements for the `compare` method since we are comparing two batches using an integer.
As for the stable order, I think it doesn't affect the current code, but I can fix this in another pr if you regard it necessary. What do you think?

Reply:
Say we have 2 batches b1 and b2 that have the same base sequence, but are not equal.  Then `compare(b1, b2) == 1` and `compare(b2, b1) == 1`, which violates the requirement of changing the sign when the argument order is changed.
This property is used in binary search tree to order and search elements, if it's violated, then we may not find the element because we follow the wrong branch.  Say we have some elements in logical order `a, b1, b2, x, y` when we ordered them we did `compare(b2, b1)`, which returned 1 meaning that b2 is greater than b1.  When we search for b1, we may start with b2 and use `compare(b1, b2)`, which would return 1 meaning that b1 is greater than b2, and we continue searching in the `x, y` part and conclude that it's not there.

Reply:
@artemlivshits Thanks for the explanation! I think it make sense. I will open another pr to fix it

Reply:
Nice catch! @artemlivshits !